Sources/Jobs/Scheduler/RecurrenceRuleConstraint.swift

```diff
@@ -0,0 +1,281 @@
+        self.validUpperBound = Calendar.gregorianUpperBound(for: timeUnit)
+
+        if stepConstraint < 1 {
+            throw RecurrenceRuleConstraintError.constraintAmountLessThanLowerBound
+        }
+        if let validUpperBound = validUpperBound {
+            if stepConstraint < 1 {
+                throw RecurrenceRuleConstraintError.constraintAmountLessThanLowerBound
+            }
+            if stepConstraint > validUpperBound {
+                throw RecurrenceRuleConstraintError.constraintAmountGreaterThanUpperBound
+            }
+        }
+
+        self.stepConstraint = stepConstraint
+    }
+
+    /// Evaluates if a given amount satisfies the constraint
+    ///
+    /// - Parameter evaluationAmount: The amount to test
+    /// - Returns: passing, failed, or noComparisonAttempted
+    public func evaluate(_ evaluationAmount: Int) -> EvaluationState {
+        // pass if evaluationAmount is divisiable of stepConstriant
+        if evaluationAmount % stepConstraint == 0 {
+            return EvaluationState.passing
+        } else {
+            return EvaluationState.failed
+        }
+    }
+
+    /// Finds the the next value that satisfies the constraint
+    ///
+    /// - Parameter currentValue: The current value the date component
+    /// - Returns: The next value that satisfies the constraint
+    public func nextValidValue(currentValue: Int) -> Int? {
+        var lowestValueGreaterThanCurrentValue: Int?
+
+        // step
+        var multiple = 0
+        var shouldStopLooking = false
+
+        if let validUpperBound = validUpperBound {
+            // others
+            var shouldStopLooking = false
+            while multiple <= validUpperBound && shouldStopLooking == false {
+                if multiple >= currentValue {
+                    if let low = lowestValueGreaterThanCurrentValue {
+                        if multiple < low {
+                            lowestValueGreaterThanCurrentValue = multiple
+                        }
+                    } else {
+                        lowestValueGreaterThanCurrentValue = multiple
+                    }
+                    shouldStopLooking = true
+                }
+                multiple = multiple + stepConstraint
+            }
+        } else {
+            // year
+            while shouldStopLooking == false {
+                if multiple >= currentValue {
+                    if let low = lowestValueGreaterThanCurrentValue {
+                        if multiple < low {
+                            lowestValueGreaterThanCurrentValue = multiple
+                        }
+                    } else {
+                        lowestValueGreaterThanCurrentValue = multiple
+                    }
+                    shouldStopLooking = true
+                }
+
+                multiple = multiple + stepConstraint
+            }
+        }
+
+        if lowestValueGreaterThanCurrentValue != nil {
```

Comment:
This can be written `return lowestValueGreaterThanCurrentValue ?? lowestPossibleValue`